### ISSUE TITLE
Memory data provider sometimes requires extra configuration

### ISSUE THREAD

**Issue by estelsmith** [drakkan/sftpgo#229]
When configuring the `memory` data provider, SFTPGo will fail to start by default on read-only filesystems, such as in an unprivileged container.

Example:

```
data_provider:
    driver: "memory"
```

Logs:

```
{"level":"debug","time":"2020-11-23T22:18:24.658","sender":"dataprovider_memory","message":"loading users from file: \"/sftpgo.db\""}
{"level":"warn","time":"2020-11-23T22:18:24.659","sender":"dataprovider_memory","message":"error loading users: stat /sftpgo.db: no such file or directory"}
{"level":"error","time":"2020-11-23T22:18:24.659","sender":"service","message":"error initializing data provider: stat /sftpgo.db: no such file or directory"}
```

---

I've found that this was being caused by the `name` configuration option having a default value, so setting `name=""` allows it to work.

It's not a huge issue, but it does cause unexpected errors that need tracked down. I'd imagine that setting the `memory` data provider would require no further configuration, as there shouldn't be a `name`d file to load data from unless I explicitly specify one.

Thoughts?

**Comment by estelsmith** [comment]
The same can be said about the `credentials_path` data provider option. It will cause SFTPGo to fail on read-only filesystems by default.

I'd think that not providing a `credentials_path` would mean that SFTPGo wouldn't attempt to touch the filesystem.

**Comment by drakkan** [comment]
Hi,

for credentials please try to set `prefer_database_credentials` to `true` this way SFTPGo should ignore the `credentials_path`.

The `name` setting allows you to configure the file that the memory provider will use to load users and folders, so we can't automatically assume it as empty, you have to configure it explicitly sorry

**Comment by estelsmith** [comment]
@drakkan Would it be preferable to default the field to empty if `memory` is chosen? That way if you specify the memory provider, it won't automatically fail in read-only environments, but still allow users to specify a file to load data from?

I feel this would strike the best balance between user-friendliness and having a default `name` set for other providers that actually require it to function.

**Comment by estelsmith** [comment]
@drakkan The same maybe could be said about the `credentials_path`? Perhaps `prefer_database_credentials` could default to `true`, only deferring to the filesystem if someone explicitly configures it to do so?

I'm not sure what's the best way for it, I was just thinking of how to make it the simplest to use without unexpected errors jumping out at the user.

**Comment by drakkan** [comment]
Hi,

should I provide a different default config file for each provider?

Actually the default configuration file works out of the box using a read/write fs and SQLite as dataprovider.

If you change the provider driver for any provider you need to change multiple settings, how can SFTPGo distinguish if you want a blank name or not? Checking if the name is `sftpgo.db` looks wrong, another user might use this name to store their dump.

Are you aware that you can change any setting using [env vars](https://github.com/drakkan/sftpgo/blob/master/docs/full-configuration.md#environment-variables)? This should simplify your docker setup.

Please note that users home dir must be writable even if you use a cloud backend.

**Comment by estelsmith** [comment]
> should I provide a different default config file for each provider?

Probably, yes. What works as a default config for sqlite may not work for memory, or other providers. It may be best to ensure default configs that rely on forcing the user to change as few config options as necessary. If it can provide a simpler default config when selecting `memory` over another provider, then should it?

It seems better than having the program unceremoniously crash on the user because the only thing they set was `data_provider.driver="memory"`.

> Actually the default configuration file works out of the box using a read/write fs and SQLite as dataprovider.

I think that's fine, and in this case the default `name` makes sense to let it know where to write the database file. I only feel it's weird to have the program crashing when setting a single configuration options.

> Are you aware that you can change any setting using env vars? This should simplify your docker setup.

Yeah, that's actually really nice. I am only using the yaml config file for testing purposes.

> Please note that users home dir must be writable even if you use a cloud backend.

Why should anything be required to be writable if I want to run it in-memory?

Basically my use-case is running SFTPGo in-memory, then using the auth and login hooks to connect it to another application providing read/write access to a cloud-backed system like S3.

**Comment by drakkan** [comment]
> > should I provide a different default config file for each provider?
> 
> Probably, yes. What works as a default config for sqlite may not work for memory, or other providers. It may be best to ensure default configs that rely on forcing the user to change as few config options as necessary. If it can provide a simpler default config when selecting `memory` over another provider, then should it?
> 
> It seems better than having the program unceremoniously crash on the user because the only thing they set was `data_provider.driver="memory"`.
> 

it refuses to start for a wrong configuration, I will evaluate your suggestion but as a first impression it seems overkill since you can easily configure it in a proper way.

Please take a look at the CI to see how each provider can be configured:

- [memory](https://github.com/drakkan/sftpgo/blob/master/.github/workflows/development.yml#L70)
- [bolt](https://github.com/drakkan/sftpgo/blob/master/.github/workflows/development.yml#L64)
- [PostgreSQL](https://github.com/drakkan/sftpgo/blob/master/.github/workflows/development.yml#L247)
- [MySQL](https://github.com/drakkan/sftpgo/blob/master/.github/workflows/development.yml#L258)

For PostgreSQL and MySQL providers is hard to have a working default configuration

> > Actually the default configuration file works out of the box using a read/write fs and SQLite as dataprovider.
> 
> I think that's fine, and in this case the default `name` makes sense to let it know where to write the database file. I only feel it's weird to have the program crashing when setting a single configuration options.
> 
> > Are you aware that you can change any setting using env vars? This should simplify your docker setup.
> 
> Yeah, that's actually really nice. I am only using the yaml config file for testing purposes.
> 
> > Please note that users home dir must be writable even if you use a cloud backend.
> 
> Why should anything be required to be writable if I want to run it in-memory?
> 

SFTPGo stores users and folders in memory but you still need a local storage for temporary files while uploading/downloading them to/from cloud backends, we use an [unlinked temporary file](https://github.com/drakkan/sftpgo/blob/master/vfs/s3fs.go#L192). This allows to use less memory (some users use 256MB part size) and at the same time to easily implement the required `WriterAt` interface. More details [here](https://pkg.go.dev/github.com/eikenb/pipeat#Pipe) 

> Basically my use-case is running SFTPGo in-memory, then using the auth and login hooks to connect it to another application providing read/write access to a cloud-backed system like S3.
> 
> Take most of this with a grain of salt. I'm only a single user

I enjoy hearing from others but cannot guarantee that I will implement every suggestion :)

**Comment by estelsmith** [comment]
> it refuses to start for a wrong configuration, I will evaluate your suggestion but as a first impression it seems overkill since you can easily configure it in a proper way.

Easy is a relative term. It's only easy for me now that I've gone through and configured the server, but at first I had no idea why the server refused to start. I don't think every data provider needs a modified default config, only the `memory` data provider should default `name` to empty, unlike the other providers.

The initially confusing part was that the [Full Configuration docs](https://github.com/drakkan/sftpgo/blob/master/docs/full-configuration.md) mentioned `name` as being optional when using the `memory` provider, but in my case it actually had to be configured, and explicitly set to blank, because I had no intention of using it.

Making `name` default to empty here would make its behavior more in-line with the docs, where it states `name` is entirely optional. To me, it indicates that `name` is expected to be empty here unless otherwise configured.

> SFTPGo stores users and folders in memory but you still need a local storage for temporary files while uploading/downloading them to/from cloud backends, we use an unlinked temporary file. This allows to use less memory (some users use 256MB part size) and at the same time to easily implement the required WriterAt interface. More details here

That makes sense, I'll keep it in mind. I probably overlooked this in the documentation.

> I enjoy hearing from others but cannot guarantee that I will implement every suggestion :)

Thanks! I'm only raising this issue as a point for brand-new users looking to try it out. This is one thing I ran across when trying it myself over the past few days. I figured out what was wrong with the config pretty quickly, thanks to your verbose logging, but found it odd that I had to troubleshoot it in the first place.

IMHO the config should require as little touching as possible to get it up and running, regardless of environment.

**Comment by drakkan** [comment]
You're right, easy is a relative term. I don't have the user's perspective :), other users have complained about the documentation and configuration in the past. Everything is obvious and easy to me since I wrote the software, the documentation definitely needs improvement. I would very much appreciate if you could submit a pull request to improve the aspects of the documentation that you found inconsistent, thank you.

Regarding the blank default name for the memory provider, I don't know exactly how to do it in code without providing a different config file which may confuse new users even more, I'll think about it

**Comment by estelsmith** [comment]
> Everything is obvious and easy to me since I wrote the software

No worries. It's just like that sometimes!

> I would very much appreciate if you could submit a pull request to improve the aspects of the documentation that you found inconsistent, thank you.

Sure! I wouldn't mind helping to improve the docs a bit.

> Regarding the blank default name for the memory provider, I don't know exactly how to do it in code without providing a different config file which may confuse new users even more, I'll think about it.

Crazy thought... Maybe the [Config struct](https://github.com/drakkan/sftpgo/blob/0609188d3f6b99851d67779ad81b13bc7ac61ffc/dataprovider/dataprovider.go#L154) can be updated to allow `nil` or `null` values in `name`, indicating it's unconfigured, not just set to `""`.

Then, in `config/config.go` there's the `init()` method that [sets the default name config](https://github.com/drakkan/sftpgo/blob/0609188d3f6b99851d67779ad81b13bc7ac61ffc/config/config.go#L118). Perhaps it could be set to `nil` or `null` by default? (I'm not super knowledgeable in Golang just yet)

After [loading the config](https://github.com/drakkan/sftpgo/blob/0609188d3f6b99851d67779ad81b13bc7ac61ffc/config/config.go#L279), you could then do the following:

- If `name` is `nil`, then check which provider was defined
- If provider is `memory`, set `name` to `""`
- Otherwise, set `name` to `"sftpgo.db"` like it currently is

I think if this were extracted into another method, it could serve as a point for modifying other default configs in the future. It's also possible I'm talking a lot of nonsense.

Thoughts??

> P.S. thank you for your sponsorship, much appreciated

☺️

**Comment by drakkan** [comment]
> > Everything is obvious and easy to me since I wrote the software
> 
> No worries. It's just like that sometimes!
> 
> > I would very much appreciate if you could submit a pull request to improve the aspects of the documentation that you found inconsistent, thank you.
> 
> Sure! I wouldn't mind helping to improve the docs a bit.
> 
thank you!

> > Regarding the blank default name for the memory provider, I don't know exactly how to do it in code without providing a different config file which may confuse new users even more, I'll think about it.
> 
> Crazy thought... Maybe the [Config struct](https://github.com/drakkan/sftpgo/blob/0609188d3f6b99851d67779ad81b13bc7ac61ffc/dataprovider/dataprovider.go#L154) can be updated to allow `nil` or `null` values in `name`, indicating it's unconfigured, not just set to `""`.
> 
> Then, in `config/config.go` there's the `init()` method that [sets the default name config](https://github.com/drakkan/sftpgo/blob/0609188d3f6b99851d67779ad81b13bc7ac61ffc/config/config.go#L118). Perhaps it could be set to `nil` or `null` by default? (I'm not super knowledgeable in Golang just yet)
> 
> After [loading the config](https://github.com/drakkan/sftpgo/blob/0609188d3f6b99851d67779ad81b13bc7ac61ffc/config/config.go#L279), you could then do the following:
> 
> * If `name` is `nil`, then check which provider was defined
> * If provider is `memory`, set `name` to `""`
> * Otherwise, set `name` to `"sftpgo.db"` like it currently is
> 
> I think if this were extracted into another method, it could serve as a point for modifying other default configs in the future. It's also possible I'm talking a lot of nonsense.
> 
> Thoughts??

to allow nil we should use a pointer, anyway the default config is updated from the config file [here](https://github.com/drakkan/sftpgo/blob/0609188d3f6b99851d67779ad81b13bc7ac61ffc/config/config.go#L269) so your solution should only work if no config file is found. I'll do some tests later or the next weekend

> 
> > P.S. thank you for your sponsorship, much appreciated

**Comment by drakkan** [comment]
Please confirm that the latest commits fixes all the raised issues, thank you for helping me to improve SFTPGo!
